pkg/build/builder/daemonless.go

```diff
@@ -311,118 +328,178 @@ func buildDaemonlessImage(sc types.SystemContext, store storage.Store, isolation
 		PullPushRetryDelay:      DefaultPushOrPullRetryDelay,
 	}
 
-	_, _, err := imagebuildah.BuildDockerfiles(opts.Context, store, options, opts.Dockerfile)
+	_, _, err = imagebuildah.BuildDockerfiles(opts.Context, store, options, opts.Dockerfile)
 	return err
 }
 
-func generateTransientMounts() []string {
-	mounts := []string{}
-	mounts = appendRHSMMount(defaultMountStart, mounts)
-	mounts = appendETCPKIMount(defaultMountStart, mounts)
-	mounts = appendRHRepoMount(defaultMountStart, mounts)
-	mounts = appendCATrustMount(mounts)
+type TransientMounts map[string]string
+
+func (t TransientMounts) append(mount string) error {
+	m := strings.Split(mount, ":")
+	if _, ok := t[m[1]]; ok {
+		return fmt.Errorf("transient mount %q already exists", m[1])
+	}
+
+	t[m[1]] = mount
+
+	return nil
+}
+
+func (t TransientMounts) asSlice() []string {
+	var mounts []string
+	for _, m := range t {
+		mounts = append(mounts, m)
+	}
 	return mounts
 }
 
-func appendRHRepoMount(pathStart string, mounts []string) []string {
+func generateTransientMounts() ([]string, error) {
+	mountsMap := make(TransientMounts)
+
+	if err := appendRHSMMount(defaultMountStart, &mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendETCPKIMount(defaultMountStart, &mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendRHRepoMount(defaultMountStart, &mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendCATrustMount(&mountsMap); err != nil {
+		return []string{}, err
+	}
+	if err := appendBuildVolumeMounts(&mountsMap); err != nil {
+		return []string{}, err
+	}
+	return mountsMap.asSlice(), nil
+}
+
+func appendBuildVolumeMounts(mountsMap *TransientMounts) error {
+	build := &buildapiv1.Build{}
+	if err := buildutil.GetBuildFromEnv(build); err != nil {
+		return err
+	}
+
+	var buildVolumeMounts []buildapiv1.BuildVolumeMount
+	switch {
+	case build.Spec.Strategy.Type == buildapiv1.SourceBuildStrategyType:
+		buildVolumeMounts = build.Spec.Strategy.SourceStrategy.VolumeMounts
+	case build.Spec.Strategy.Type == buildapiv1.DockerBuildStrategyType:
+		buildVolumeMounts = build.Spec.Strategy.DockerStrategy.VolumeMounts
+	}
+
+	for _, bvm := range buildVolumeMounts {
+		var mountOptions []string
+		if bvm.ReadOnly {
+			mountOptions = append(mountOptions, "ro")
+		} else {
+			mountOptions = append(mountOptions, "rw")
+		}
+
+		if err := mountsMap.append(fmt.Sprintf("%s:%s:%s", filepath.Join(BuildVolumeMountPath, bvm.MountPath), bvm.MountPath, strings.Join(mountOptions, ","))); err != nil {
+			return err
+		}
```

Comment:
If `append` accepts a fuller TransientMount object as an argument, then this logic is not repeated.